yarn-project/prover-client/src/proving_broker/proving_broker.ts

```diff
@@ -0,0 +1,341 @@
+import {
+  ProvingRequestType,
+  type V2ProvingJob,
+  type V2ProvingJobId,
+  type V2ProvingJobResult,
+  type V2ProvingJobStatus,
+  type V2ProvingResult,
+} from '@aztec/circuit-types';
+import { createDebugLogger } from '@aztec/foundation/log';
+import { RunningPromise } from '@aztec/foundation/promise';
+import { PriorityMemoryQueue } from '@aztec/foundation/queue';
+
+import assert from 'assert';
+
+import { type ProvingBrokerDatabase } from './proving_broker_database.js';
+import type { ProvingJobConsumer, ProvingJobFilter, ProvingJobProducer } from './proving_broker_interface.js';
+
+type InProgressMetadata<T extends ProvingRequestType = ProvingRequestType> = {
+  id: V2ProvingJobId<T>;
+  startedAt: number;
+  lastUpdatedAt: number;
+};
+
+type ProofRequestBrokerConfig = {
+  timeoutIntervalMs?: number;
+  jobTimeoutMs?: number;
+  maxRetries?: number;
+};
+
+/**
+ * A broker that manages proof requests and distributes them to workers based on their priority.
+ * It takes a backend that is responsible for storing and retrieving proof requests and results.
+ */
+export class ProvingBroker implements ProvingJobProducer, ProvingJobConsumer {
+  // Each proof type gets its own queue so that agents can request specific types of proofs
+  // Manually create the queues to get type checking
+  private queues: ProvingQueues = {
+    [ProvingRequestType.PUBLIC_VM]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.TUBE_PROOF]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.PRIVATE_KERNEL_EMPTY]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+
+    [ProvingRequestType.PRIVATE_BASE_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.PUBLIC_BASE_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.MERGE_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.ROOT_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+
+    [ProvingRequestType.BLOCK_MERGE_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.BLOCK_ROOT_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.EMPTY_BLOCK_ROOT_ROLLUP]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+
+    [ProvingRequestType.BASE_PARITY]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+    [ProvingRequestType.ROOT_PARITY]: new PriorityMemoryQueue<V2ProvingJob>(provingJobComparator),
+  };
+
+  // holds a copy of the database in memory in order to quickly fulfill requests
+  // this is fine because this broker is the only one that can modify the database
+  private jobsCache = new Map<V2ProvingJobId, V2ProvingJob>();
+  // as above, but for results
+  private resultsCache = new Map<V2ProvingJobId, V2ProvingJobResult>();
```

Comment:
Given lmdb is already in-memory, is there a significant improvement by not having to cross to the C-land boundary and keeping these copies here?